datajoint/external.py

```diff
@@ -6,6 +6,203 @@
+                .format(store=store))
+
+        protocol = spec['protocol']
+
+        if protocol not in cls._handlers:
+            raise DataJointError(
+                'Unknown external storage protocol "{protocol}" for "{store}"'
+                .format(store=store, protocol=protocol))
+
+        return cls._handlers[protocol](store, database)
+
+    def check_required(self, store, storetype, required):
+
+        missing = list(i for i in required if i not in self._spec)
+
+        if len(missing):
+            raise DataJointError(
+                'Store "{s}" incorrectly configured for "{t}", missing: {m}'
+                .format(s=store, t=storetype, m=missing))
+
+    def hash_obj(self, obj):
+        blob = pack(obj)
+        hash = long_hash(blob) + self._store[len('external-'):]
+        return (blob, hash)
+
+    @staticmethod
+    def hash_to_store(hash):
+        store = hash[STORE_HASH_LENGTH:]
+        return 'external' + ('-' if store else '') + store
+
+    def put(self, obj):
+        ''' returns (blob, hash) '''
+        raise NotImplementedError(
+            'put method not implemented for', type(self))
+
+    def get_blob(self, hash):
+        ''' returns undecoded 'blob' '''
+        raise NotImplementedError(
+            'get_blob method not implemented for', type(self))
+
+    def get_obj(self, hash):
+        ''' returns decoded 'obj' '''
+        raise NotImplementedError(
+            'get_obj method not implemented for', type(self))
+
+    def get(self, hash):
+        ''' returns decoded 'obj' '''
+        return self.get_obj(hash)
+
+
+class RawFileHandler(ExternalFileHandler):
+
+    required = ('location',)
+
+    def __init__(self, store, database):
+        super().__init__(store, database)
+        self.check_required(store, 's3', RawFileHandler.required)
+        self._location = self._spec['location']
+
+    def get_folder(self):
+        return os.path.join(self._location, self._database)
+
+    def put(self, obj):
+        (blob, hash) = self.hash_obj(obj)
+
+        folder = self.get_folder()
+        full_path = os.path.join(folder, hash)
+        if not os.path.isfile(full_path):
+            try:
+                with open(full_path, 'wb') as f:
+                    f.write(blob)
+            except FileNotFoundError:
+                os.makedirs(folder)
+                with open(full_path, 'wb') as f:
+                    f.write(blob)
+
+        return (blob, hash)
+
+    def get_blob(self, hash):
+        full_path = os.path.join(self.get_folder(), hash)
+        try:
+            with open(full_path, 'rb') as f:
+                return f.read()
+        except FileNotFoundError:
+                raise DataJointError('Lost external blob')
+
+    def get_obj(self, hash):
+        return unpack(self.get_blob(hash))
+
+    def get(self, hash):
+        return self.get_obj(hash)
+
+
+class CacheFileHandler(ExternalFileHandler):
```

Comment:
I don't think cache is its own store type.  Rather, it's a feature in all other store types.

Reply:
Or we could do cache in the `ExternalTable.put`/`get` and avoid problems with inheritance.